### PR TITLE
DeepChem learns to play tictactoe

### PR DESCRIPTION
**Pull request by lilleswing**
I played around with a3c and RL in Deepchem seeing if I could train a tictactoe agent.

The agent isn't good yet and is still having issues.

1) after training for a long time the Dense networks output NaN instead of float32 values
2) The agent just makes illegal moves constantly.

I still wanted to put this  up for CR because I found some small things in A3C that might trip up others so I wanted to fix them.

1) Deep Copy the environment's state in the worker thread.  Environments will most likely mutate their environment on step commands, without this all states would just be the final state for the rollout.

2) Ininitalize the global variables after the worker threads are created.  This might be wrong @peastman .  This was needed because workers were creating their own variables on worker thread creation.  How did you get it to work otherwise?  Do we want the layers shared between the workers or each worker to have their own network?

### CODE REVIEW

**Comment by peastman** [comment]
Thanks, I'll take a look!

File: deepchem/rl/a3c.py [hunk 1]
```diff
@@ -237,7 +237,7 @@ def __init__(self, a3c, index):
       self.train_op = a3c._graph._get_tf('Optimizer').apply_gradients(
           grads_and_vars)
       self.update_local_variables = tf.group(
-          * [tf.assign(v1, v2) for v1, v2 in zip(local_vars, global_vars)])
+          *[tf.assign(v1, v2) for v1, v2 in zip(local_vars, global_vars)])
```
Comment (peastman) [inline deepchem/rl/a3c.py]: yapf is complaining about this change.  It's one of those super irritating cases where yapf produces a different result on some computers than others, so if you run it locally, it produces formatting that travis rejects as wrong.

File: deepchem/rl/a3c.py [hunk 2]
```diff
@@ -263,7 +263,7 @@ def create_rollout(self):
     for i in range(self.a3c.max_rollout_length):
       if self.env.terminated:
         break
-      state = self.env.state
+      state = copy.deepcopy(self.env.state)
```
Comment (peastman) [inline deepchem/rl/a3c.py]: This shouldn't be needed.  States should be immutable.  The problem is that you've implemented your environment so that it modifies states after outside code has already retrieved them.  Doing that opens the door to all sorts of bugs in lots of places.  Instead it should assign a new value to `_state` which is then treated as immutable.

**Comment by peastman** [comment]
> How did you get it to work otherwise?

No idea!  Your change looks correct though.

> Do we want the layers shared between the workers or each worker to have their own network?

Each worker needs its own network.  This is actually one of the things that makes A3C more robust: different workers may be using slightly different networks at the same time, which increases the diversity of trajectories they explore.

File: deepchem/rl/a3c.py (hunk 2 above)
Comment (lilleswing) [inline deepchem/rl/a3c.py]: While I understand that decision -- I don't think it will be clear to most users and will lead to users mutating state in the Environment with step() and reset().  We should at least document it on the Environment object, state property, and step function.

Doing it here is safer and easier to reason about -- but will lead to worse performance in the case of someone actually making their state immutable, and creating a new one each time.

Comment (lilleswing) [inline deepchem/rl/a3c.py]: I can change my tictactoe board and document the functions though if we decide to ask for immutability of state by the env.

Comment (peastman) [inline deepchem/rl/a3c.py]: We should definitely document it for people writing environments.  It's not just this one line—it would create opportunities for bugs all over the place.  Any piece of code that ever stores a state would start having to worry that state might magically change behind its back.

**Comment by lilleswing** [comment]
The NaNs were from tf.log(prob) where prob was a probability vector.  A single value which == tf.float32(0) would destroy the whole network.  I added an epsilon to get around this.

**Comment by rbharath** [comment]
I really like the idea of adding more environments to DeepChem. New environments are the only real way of benchmarking our RL suite, and it would be great to build up our capabilities. In fact, I have a couple extra suggestions:

- Can you make a new submodule `dc.rl.envs` that hosts environments and move the Tic-tac-toe environment there?
- Can you add a section to the benchmarks on the `README.md` that documents best performance on the Tic-Tac-Toe environment? I think we should start keeping running numbers on the RL environments like we do for MoleculeNet (betting RL is going to be increasingly important moving forward).

**Comment by peastman** [comment]
If you want lots of environments, you could just make OpenAI Gym an optional dependency.  We've already got support for it via the GymEnvironment class, so no new code is even required.  Just start adding benchmarks or examples that use it.

**Comment by rbharath** [comment]
Definitely like the idea of making OpenAI gym an optional dependency. Any PRs adding benchmarks or examples in this vein welcome.

File: deepchem/rl/a3c.py [hunk 3]
```diff
@@ -21,12 +21,13 @@ def __init__(self, value_weight, entropy_weight, **kwargs):
     self.entropy_weight = entropy_weight
 
   def create_tensor(self, **kwargs):
-    reward, action, prob, value = [layer.out_tensor for layer in self.in_layers]
-    log_prob = tf.log(prob)
+    reward, action, prob, value, advantages = [layer.out_tensor for layer in self.in_layers]
+    log_prob = tf.log(prob + 0.0001)
     policy_loss = -tf.reduce_sum(
```
Comment (lilleswing) [inline deepchem/rl/a3c.py]: So isn't this just another value loss?  What is encouraging the network to seed out better rewards/values?  

The way I am reading this loss function right now is it is encouraging certainty -- that is the value loss, policy loss both encourage the value==discounted_rewards.  The entropy also encourages certainty, to not be unsure but to go where you know where you have accurate value function.  

I implemented advantage as seen here
https://medium.com/emergent-future/simple-reinforcement-learning-with-tensorflow-part-8-asynchronous-actor-critic-agents-a3c-c88f72a5e9f2
In order to make the policy loss portion encourage going towards higher rewards.

Still not able to play tic tac toe though.

Am I reading the current implementation of policy loss wrong @peastman ?

File: deepchem/rl/a3c.py [hunk 4]
```diff
@@ -21,10 +21,11 @@ def __init__(self, value_weight, entropy_weight, **kwargs):
...
   def create_tensor(self, **kwargs):
-    reward, action, prob, value = [layer.out_tensor for layer in self.in_layers]
-    log_prob = tf.log(prob)
-    policy_loss = -tf.reduce_sum(
-        (reward - value) * tf.reduce_sum(action * log_prob))
+    reward, action, prob, value, advantages = [
+        layer.out_tensor for layer in self.in_layers
+    ]
+    log_prob = tf.log(prob + 0.0001)
+    policy_loss = -tf.reduce_sum(advantages * tf.reduce_sum(action * log_prob))
```
Comment (peastman) [inline deepchem/rl/a3c.py]: I don't understand what you're doing here.  Advantage is defined as reward-value, where the reward is the actual reward that was obtained in the rollout, and value is the value estimate produced by the policy.  That's precisely what it was already using.

Comment (lilleswing) [inline deepchem/rl/a3c.py]: I don't think that is the definition of advantage.
From the A3C paper
https://arxiv.org/pdf/1602.01783.pdf

selection_094

What I implemented here is k=1

Comment (lilleswing) [inline deepchem/rl/a3c.py]: Page 4 bottom left corner

**Comment by lilleswing** [comment]
@rbharath what do we mean by performance?  

* avg score of a rollout?
  * How much do we allow to train?
* Time to a given avg score?
* Max score of a rollout?

Performance for these things is *hard* because everyone disagrees on what performance means.

**Comment by rbharath** [comment]
On my phone, so will respond about perf in more depth later. In meantime,
this new effort by OpenAI is very pertinent

https://blog.openai.com/openai-baselines-dqn/amp/

File: deepchem/rl/a3c.py (hunk 4 above)
Comment (peastman) [inline deepchem/rl/a3c.py]: That isn't necessarily the correct equation to use.  They say the update "can be seen as" an estimate of that function.  However, the actual loss function they use appears two paragraphs later.  It just depends on `R-V` without that extra term.

Also see https://arxiv.org/pdf/1506.02438.pdf for a much more detailed discussion of advantage estimation.  They present both forms (eq. 17 and 18) as limiting cases of a continuous family of advantage estimates (eq. 16).  The one I implemented (`lambda=1`) is unbiased but has larger variance.  The one you used (`lambda=0`) reduces the variance but introduces bias.  In their experiments, they get the best results for values in the range of about 0.96, so much closer to 1 than 0.  We could consider adding `lambda` as a user selectable hyperparameter.

Comment (lilleswing) [inline deepchem/rl/a3c.py]: Ok I'll update A3C to have lambda as a hyper parameter.  Without the change in advantage my agent does not learn to play tictactoe very well.

I'm interested what the grid of avg reward looks like for my network for differing lambda and gamma.  After I implement lambda as a hyper param I'll update with my grid of avg reward along with the code.

Comment (peastman) [inline deepchem/rl/a3c.py]: Sounds good.

Comment (peastman) [inline deepchem/rl/a3c.py]: I'd also suggest experimenting with different values for `value_weight`.  It's possible the problem is just that your weight is too low, so it's taking a long time to learn the value function.

Also, are you sure it's converged?  100,000 steps isn't very many.

Comment (peastman) [inline deepchem/rl/a3c.py]: Try going up to much larger values for `value_weight`.  A possible explanation for what you see is just that it isn't learning the value function (because you've set the weight too low).

Comment (peastman) [inline deepchem/rl/a3c.py]: You also could try directly querying the value function for various states, to see whether it's learned it or not.

Comment (lilleswing) [inline deepchem/rl/a3c.py]: @peastman you were right there was a value_weight sweat spot between 4-6.

selection_003

selection_004

The model isn't as stable as with lambda=0 but if you use a pocket algorithm it can produce similar results.

Comment (peastman) [inline deepchem/rl/a3c.py]: Good to know.  It will be interesting to see the full matrix of how it depends on gamma, lambda, and value_weight.

File: deepchem/rl/a3c.py [hunk 5]
```diff
@@ -22,7 +22,7 @@ def __init__(self, value_weight, entropy_weight, **kwargs):
 
   def create_tensor(self, **kwargs):
     reward, action, prob, value = [layer.out_tensor for layer in self.in_layers]
-    log_prob = tf.log(prob)
+    log_prob = tf.log(prob + 0.0001)
```
Comment (peastman) [inline deepchem/rl/a3c.py]: Why 0.0001?  If the goal is just to make sure we never take the log of zero, then machine epsilon seems like a better value.  You can get it as `np.finfo(np.float32).eps`.  Also, might it be better to use `max(prob, eps)` rather than addition?  That will distort the probability distribution less.

**Comment by peastman** [comment]
Aside from the one comment I just made about the additive constant, this looks good to me.

**Comment by peastman** [comment]
By the way, after looking at the equation you highlighted in the paper again I realized I was misinterpreting it.  The extra term is nothing more than adding a single V term to estimate everything beyond the end of the rollout.  We already include that.

https://github.com/deepchem/deepchem/blob/master/deepchem/rl/a3c.py#L276-L280

**Comment by lilleswing** [comment]
@peastman huh I misread that chunk of code assuming it was doing some sort of simulation for the rest of a rollout if it ended early.

That makes the improved performance after my changes to advantage even more unexplained.

**Comment by peastman** [comment]
It's only one term, not a sum of terms.  Remember, V is the *total* expected return starting from that state, not the expected return from just one step.  If the task has terminated, it's zero by definition (there will be no more rewards), so there's nothing to add.  But if you ended the task early because you reached the maximum rollout length, you want to add an estimate of the additional reward you would have gotten if you hadn't stopped there.

**Comment by lilleswing** [comment]
I'm going to summarize what I learned here for the github issue.  I didn't realize code lines that were overwritten causes the comments to get lost.

When lambda=0 for the given state you explicitly add the difference in value function between the two consecutive states as part of the policy loss.  This is effectively leaning as hard as you can on short term gain.

When lambda=1 this is favoring long term gain.  A "bad move" at the very end of your game gets discounted through the whole rollout and used as the "score" of every policy step.

I suspect why lambda=0 works for tictactoe is the main issue is making illegal moves late in the game.  An illegal move late in the game does not necessarily mean every move before has been bad -- but the discounted reward for each state up to that point will be a poor score.

https://arxiv.org/pdf/1506.02438.pdf is a paper which goes more in depth about Generalized Advantage Estimation (Michael Jordan and Pieter Abbeel).

File: deepchem/rl/a3c.py (hunk 5 above)
Comment (lilleswing) [inline deepchem/rl/a3c.py]: I added np.finfo(np.float32).eps.  I'm willing to take the hit into the distortion of probability distribution for the added simplicity.

**Comment by peastman** [comment]
Not quite.  The difference isn't a matter of favoring short term vs. long term gain.  (That's what gamma is for.)  The difference is how you estimate long term gain.  There are two ways you might try to estimate the value function given the state at step i: `V(s_i) = r_i + r_(i+1) + r_(i+2) + ...` or `V(s_i) = r_i + V(s_(i+1))`.  (For simplicity I'm assuming gamma is 1.)  The first one is unbiased, but has a high variance since it's a sum of many terms.  The second one is much lower variance, since it only depends on one step, but it assumes you already know how to calculate the value function for the next step.  But that's the quantity we're trying to learn, so all we can do is use our current estimate of it, and that introduces bias.  The purpose of lambda is to interpolate between these two approximations.

As for why you get better results with lambda=0, I'd say it's because RL is super finicky and tiny changes can have big effects on the results, often not for any obvious reason. :)

**Comment by rbharath** [comment]
Looked through the code and things look good to me. The TicTacToe seems like a simple, but quite nontrivial learning problem, so will be a good test bed. Looking forward to seeing how DQN style methods do on it.

LGTM.

**Comment by rbharath** [comment]
Apparently pull requests with a failing status can't be merged on a phone... Will merge in once back at desk.

**Comment by rbharath** [comment]
Looks like there's a yapf failure. Good to merge in once that's fixed.

**Comment by lilleswing** [comment]
@rbharath @peastman sorry I was on vacation and out of internet contact for the past 1.5 weeks.  I fixed the YAPF problem and this is good to go now.

**Comment by rbharath** [comment]
@lilleswing Awesome, thanks for the fix. LGTM.
